dashboard/data/structural_similarity.py

```diff
@@ -0,0 +1,202 @@
+    merged = smiles_active.merge(activity, on="EOS")
+    smiles_new["activity_final"] = "not tested"
+    smiles_new = smiles_new.rename(columns={"eos": "EOS"})
+    merged = pd.concat([merged, smiles_new])
+    merged = merged.drop_duplicates(subset=["EOS"])
+    return merged
+
+
+def calculate_umap(descriptors: np.ndarray) -> np.ndarray:
+    """
+    Calculate UMAP projection
+
+    :param descriptors: descriptors
+    :return: array with projection
+    """
+    umap_model = umap.UMAP(
+        metric="jaccard",
+        n_neighbors=25,
+        n_components=2,
+        low_memory=False,
+        min_dist=0.001,
+    )
+    X_umap = umap_model.fit_transform(descriptors)
+    return X_umap
+
+
+def calculate_pca(descriptors: np.ndarray) -> np.ndarray:
+    """
+    Calculate PCA projection
+
+    :param descriptors: descriptors
+    :return: array with projection
+    """
+    pca_model = PCA(n_components=2)
+    X_pca = pca_model.fit_transform(descriptors)
+    return X_pca
+
+
+def calculate_clusters(x_projection: np.ndarray) -> np.ndarray:
+    """
+    Calculate hdbscan clusters
+
+    :param x_projection: array with projected data
+    :return: array with clusters
+    """
+    hdbscan_model = hdbscan.HDBSCAN(
+        min_cluster_size=20, min_samples=20, cluster_selection_method="eom"
+    )
+    preds = hdbscan_model.fit_predict(x_projection)
+    clusters = [f"c{x}" if x != -1 else "outlier" for x in preds]
+    return clusters
+
+
+def prepare_cluster_viz(
+    activity: pd.DataFrame, smiles_active: pd.DataFrame, smiles_new: pd.DataFrame
+) -> pd.DataFrame:
+    """
+    Merge dataframes, calculate projections and clusters
+
+    :param activity: df with activity calculated
+    :param smiles_active: df with smiles of active compounds
+    :param smiles_new: new smiles to cluster
+    :return: df with everything calculated
+    """
+    df = merge_active_new(activity, smiles_active, smiles_new)
+    ecfp_descriptors, keep_idx = compute_ecfp_descriptors(df["smiles"])
+    df = df.iloc[keep_idx]
+    X_umap = calculate_umap(ecfp_descriptors)
+    df["UMAP_0"], df["UMAP_1"] = X_umap[:, 0], X_umap[:, 1]
+    df["cluster_UMAP"] = calculate_clusters(X_umap)
+    X_pca = calculate_pca(ecfp_descriptors)
+    df["PCA_0"], df["PCA_1"] = X_pca[:, 0], X_pca[:, 1]
+    df["cluster_PCA"] = calculate_clusters(X_pca)
+    return df
+
+
+def plot_clustered_smiles(
+    df: pd.DataFrame,
+    feature: str = "activity_final",
+    projection: str = "PCA",
+) -> go.Figure:
+    """
+    Plot selected projection and colour points with respect to selected feature.
+
+    :param df: DataFrame to be visualized
+    :param feature: name of the column with respect to which the plot will be coloured
+    :param projection: name of projection to be visualized
+
+    :return: plotly express scatter plot
+    """
+    projection_x = f"{projection.upper()}_0"
+    projection_y = f"{projection.upper()}_1"
+    clusters = f"cluster_{projection}"
+    labels = {
```

Comment:
perhaps this could be moved to `plots.py` where we have all functions for plotly charts generation (and configuration)